### DISCUSSION TITLE
Research QML as substitute for Layers

### DISCUSSION THREAD

**Discussion by joewheaton** [Riverscapes/RiverscapesXML#180] [RAVE: Symbology/BusinessLogic] [answered]
## The Problem
RAVE currently uses `*.lyr` files as basis for building its symbology that are from ArcGIS 10.4.  The problem is that none of the development team (other than @philipbaileynar) use ArcGIS 10.4 and cannot produce old Layer files without uninstalling their current version or finding an old 10.4 version on another machines. The other irony is this open-source piece of software using something horrible and proprietary like `*.lyr`.

## What we found quickly.
1. I did find out that ESRI has [this list of what functionality](desktop.arcgis.com/en/arcmap/latest/map/working-with-arcmap/saving-a-map.htm#ESRI_SECTION1_7621A3BAFFED4A809BA7065228BFDF62) is actually different between versions of ArcGIS and layer files.
2. We also learned about [QML](doc.qt.io/archives/qt-4.8/qdeclarativebasictypes.html), which apparently can be exported from both [QGIS](https://docs.qgis.org/2.8/en/docs/training_manual/basic_map/symbology.html) and ArcGIS and 
3. There is a even a [batch QML tool for Q](https://anitagraser.com/2012/02/17/batch-application-of-qgis-layer-styles/).

## Work Around for Now

For Cashe and anyone for now, we can produce 10.4 layer files by just doing it at the Python command line. Thankfully the '[Save to Layer File](https://desktop.arcgis.com/en/arcmap/latest/tools/data-management-toolbox/save-to-layer-file.htm)'  command has an optional argument for the version of ArcGIS for the Output layer file! 

![image](https://user-images.githubusercontent.com/9723905/81023474-c354ab80-8e2d-11ea-86bb-8dea4d80c680.png)

So just open a Python window in your current ArcGIS session and type following 
```
>>> import arcpy
>>> arcpy.env.workspace = "C:/0_GIS"
>>> arcpy.SaveToLayerFile_management(r"VBET for HUC 14050001\Inputs\Topography\NHD Flowlines","C:/0_GIS/NHD_Flowlines.lyr","ABSOLUTE","10.4")
<Result 'C:\\0_GIS\\NHD_Flowlines.lyr'>
```

Obviously substitute whatever you want above.

# Long-Term Solution
We think the long-term solution is to switch over to adopting QML and this might have benefits for going to web... @philipbaileynar  will research.

**Comment by joewheaton** [comment]
BTW... There is also a toolbox function if you don't like command prompt @wally-mac:
![image](https://user-images.githubusercontent.com/9723905/81024003-4b878080-8e2f-11ea-95ac-b51c886bae30.png)

**Comment by philipbaileynar** [comment]
I just posted this question to GIS Stack Exchange:

https://gis.stackexchange.com/questions/360591/can-arcgis-desktop-read-any-open-source-layer-definitions

**Comment by joewheaton** [comment, marked as answer]
This is working now in QRAVE. Also, see #152,  https://github.com/Riverscapes/RaveAddIn/issues/138.
